sqlmesh/core/model/definition.py

```diff
@@ -2455,3 +2469,43 @@ def _meta_renderer(
 def get_model_name(path: Path) -> str:
     path_parts = list(path.parts[path.parts.index("models") + 1 : -1]) + [path.stem]
     return ".".join(path_parts[-3:])
+
+
+# function applied to time column when automatically used for partitioning in INCREMENTAL_BY_TIME_RANGE models
+def clickhouse_partition_func(
+    column: exp.Expression, columns_to_types: t.Dict[str, exp.DataType]
+) -> exp.Expression:
+    # `toMonday()` function accepts a Date or DateTime type column
+
+    col_type_present = columns_to_types and column.name in columns_to_types
+    col_type_is_conformable = col_type_present and columns_to_types[column.name].is_type(
+        exp.DataType.Type.DATE,
+        exp.DataType.Type.DATE32,
+        exp.DataType.Type.DATETIME,
+        exp.DataType.Type.DATETIME64,
+    )
+
+    #  if input column is already a conformable type, just pass the column
+    if col_type_is_conformable:
+        return exp.func("toMonday", column)
```

Comment:
I'd do `dialect="clickhouse"` here just to avoid any dialect-specific parsing shenanigans.